Varieties v0.3

mahrud opened the pull request:
This PR contains work throughout Spring and Summer 2024, including at the workshop at Utah. Complexes of sheaves are added, but are currently disabled until the Complexes package can be preloaded. Lots of work remains.

cc: @kellerlv @ggsmith @Devlin-Mallory @ritvikrr @johndcobb

mahrud commented:
I'm not sure if Mike has forgotten about https://github.com/mikestillman/M2/pull/38, but this PR also includes those changes.

mahrud commented:
@d-torrance any idea what's happening here? I don't think this PR changes anything related to this ...
```m2
--*- compilation -*-

i1 : input "../../../../../Macaulay2/tests/normal/alarm.m2"

ii2 : -- no-check-flag (see https://github.com/Macaulay2/M2/issues/1392)
      -- test that alarms work
      
      time try ( alarm(1); while true do 1 ) else true
 -- used 1.13874s (cpu); 0.999088s (thread); 0s (gc)

oo2 = true

ii3 : 
      R = QQ[vars(0..24)]

oo3 = R

oo3 : PolynomialRing

ii4 : f = () -> (alarm 4; try res coker vars R else "ran out of time")

oo4 = f

oo4 : FunctionClosure

ii5 : time f()


 *** out of memory trying to allocate 128 bytes, exiting ***
Command exited with non-zero status 1
16.95user 4.55system 0:11.87elapsed 181%CPU (0avgtext+0avgdata 477640maxresident)k
0inputs+8outputs (0major+13659minor)pagefaults 0swaps
```

d-torrance commented:
Not sure.  Apparently this test has been a problem before since we actually skip it when when running `check Core` (see #1867).

Devlin-Mallory commented on M2/Macaulay2/packages/Varieties.m2:
```diff
@@ -159,7 +162,8 @@ isWellDefined Variety := X -> (
-    and assert'(not isProjective X or isHomogeneous R,
+    -- TODO: support non-graded rings (e.g. ZZ, QQ, etc.) and nonstandard gradings
+    and assert'(not isProjective X or isStandardGraded R,
 	"coordinate ring of a projective variety should be homogeneous")
```
perhaps the complaint should read "should be homogeneous and generated in degree 1"

d-torrance commented:
@mahrud - Do you think it's worth applying the `SumsOfSquares` test fix on `development` separately while we wait for this PR to be review?

mahrud commented:
> Do you think it's worth applying the `SumsOfSquares` test fix on `development` separately while we wait for this PR to be review?

I think you meant to comment this on #3519, where bbc9f09 is, no? I think it's good to keep it there as extra incentive for reviewing it.

d-torrance commented:
Sure -- it looks like this PR is on top of that one, so either way.  :)  Sounds good!

ggsmith commented on M2/Macaulay2/m2/modules2.m2:
```diff
@@ -326,27 +327,32 @@ Module ^ List := Matrix => (M, rows) -> submatrix(map(cover M, M, id_M), rows,)
-	inducedMap(P, T) * map(T, target g, T_[1], Degree => - degree g)};
+    P.cache.formation = FunctionApplication (pushout, args);
+    P.cache.pushoutMaps = apply(#args,
+	i -> inducedMap(P, T) * map(T, target args#i, T_[i], Degree => - degree args#i));
```
Are the ``pullback`` and ``pushout`` documented or tested anywhere?

mahrud replied:
They are tested in `Varieties`, but the documentation is only a stump.

ggsmith commented on M2/Macaulay2/packages/Book3264Examples.m2:
```diff
@@ -64,7 +64,7 @@ placeholderToSchubertBasis(RingElement,FlagBundle) := (c,G) -> (
      P := diagrams(q,k);
      M := apply(P, i-> placeholderSchubertCycle(i,G));
-     E := flatten entries basis(R);
+     E := flatten entries basis(R, Variables => 0 .. numgens R - 1);
```
This appears to depend on other changes to ``basis``, which I am personally not yet sold on.

ggsmith commented on M2/Macaulay2/packages/Complexes/CartanEilenberg.m2:
```diff
@@ -0,0 +1,420 @@
+
+--not yet written
+-*
+doubleHorseshoe = method()
```
I believe that the ``SpectralSequences`` package has enjoy a major revision which I suspect is incompatible with this code.

mahrud replied:
Oh I didn't realize this file (which isn't actually loaded) was put in Complexes. I will exclude the commit that adds this for now.

ggsmith commented on M2/Macaulay2/m2/basis.m2:
```diff
@@ -183,12 +197,47 @@ basis(List,           List,           Matrix) :=
+    -- TODO: should matrix RingMap return this map instead?
+    -- mon := map(module T, image matrix(S, { S.FlatMonoid_* }), f, matrix f)
+    mat := last coefficients(f cover srcgens, Monomials => targens);
+    map(image targens, image srcgens, f, mat))
```
Aren't these many changes to ``basis`` already part of an independent pull request?

mahrud replied:
Yes, but this PR is on top of that because of actual dependencies (but it is _not_ on top of the `part` PR).

ggsmith commented on M2/Macaulay2/m2/matrix2.m2:
```diff
@@ -325,32 +325,40 @@ quotientRemainder(Matrix,Matrix) := Matrix => (f,g) -> (
+Matrix \\ Matrix := Matrix => (g, f) -> quotient'(f, g)
+-- commented because they don't seem very meaningful
+--Matrix \\ Number := Matrix \\ RingElement := Matrix => (g, r) -> map(source g, target g, g \\ map(cover source g, source g, r))
+--Number \\ Matrix := RingElement \\ Matrix := Matrix => (r, f) -> map(target g, source g, map(cover source g, source g, r) \\ f)
```
Can we make the (breaking?) changes to quotients of ``Matrices`` a separate pull request?  I was unable to find the associated documentation and tests.

mahrud replied:
I tried to do this and failed. I don't think it's impossible, but I decided it wasn't worth the time and effort to essentially redo everything in a different order.

ggsmith commented on M2/Macaulay2/packages/Isomorphism.m2:
```diff
@@ -193,8 +193,7 @@ isIsomorphic(Module, Module) := Sequence => o ->  (N,M)->(
     )
-isIsomorphic(Matrix,Matrix) := Sequence => o -> (n,m) -> 
-           isIsomorphic(coker m, coker n)
+isIsomorphic(Matrix, Matrix) := Sequence => o -> (n, m) -> isIsomorphic(coker m, coker n, o)
```
- [x] good catch

ggsmith commented on M2/Macaulay2/packages/Complexes/ChainComplex.m2:
```diff
@@ -1154,7 +1162,7 @@ yonedaExtension Matrix := Complex => f -> (
     -- if g has a non-zero degree, we must twist the target to preserve homogeneity
     gdegree := degree g;
-    g = map(N ** (ring g)^gdegree, source g, g);
+    g = map(N ** (ring g)^{gdegree}, source g, g);
```
- [x] good idea

mahrud replied:
Thanks!

ggsmith commented on M2/Macaulay2/packages/Complexes/ChainComplexMap.m2:
```diff
@@ -99,13 +99,9 @@ map(Complex, Complex, ZZ) := ComplexMap => opts -> (D, C, j) -> (
-    )
+    if C == D and (opts.Degree === null or opts.Degree === 0) then
+        return j * id_C;
+    error "expected 0 or source and target to be the same")
```
I think that this change is already part of a pull request associated to ``Complexes``.

mahrud replied:
Yes, it's the same commit, so once one of them is merged, the duplicate commit will be skipped in the other PR.

ggsmith commented on M2/Macaulay2/packages/NormalToricVarieties/Divisors.m2:
```diff
@@ -177,7 +177,7 @@ nefGenerators NormalToricVariety := Matrix => X -> (
     indexOfPic := abs lcm (minors( rank source fromPic, fromPic^torsionlessCoord))_*;
     (indexOfPic * coneGens) // fromPic 
-    );
+    ))
```
By definition, the nef cone lives in the Neron--Severi group (which for a toric variety is canonical isomorphic to the Picard group).  Does returning something in the class group make mathematical sense?

mahrud replied:
Sorry, I don't follow. The only change here is caching of `nefGenerators`. What change are you referring to?

ggsmith reviewed (changes requested):
Can this pull request be decoupled from the changes to ``basis``?  Can the changes to quotient modules and matrices to extracted?

mahrud commented:
> Can this pull request be decoupled from the changes to `basis`?

It is decoupled. I don't remember exactly why, but the changes to basis stemmed from fixing things in Varieties, which is why this PR is on top of that one.

> Can the changes to quotient modules and matrices to extracted?

I tried to do this a while ago but I'm afraid it was very complicated (random tests kept failing in one package or another when I tried to separate them) and I gave up, so I don't want to do it. If you or someone else wants to give it an attempt you're welcome to.

mahrud commented:
Okay, updates:
- Opened #3550 for quotient changes (though this PR is still on top of it), I think the errors I couldn't fix last time were SumsOfSquares errors, I don't remember.
- Fixed the error message for non-standard graded rings
- Removed the Cartan-Eilenberg commit (it is safe on another branch for later)

mikestillman commented on M2/Macaulay2/packages/Varieties/backports.m2:
```diff
@@ -0,0 +1,3 @@
+
```
What changes are in this file?

mahrud replied:
Ah sorry this was used during the workshop, should have been removed in https://github.com/Macaulay2/M2/commit/d90df86a02f09eb22644644af8db8fa932949579. I'll do it in the next push.

mikestillman reviewed (commented):
Once @ggsmith is happy with this, it looks fine to me, with the caveat that I'm not so convinced that the breaking change to basis is worth the breaking.  I think `basis` needs a complete revamping at some point, with more expressive names for what it is doing.  But other than that, this all looks good.  I did not go into all the Varieties code in detail though.  I am happy to see complexes of sheaves though!

mahrud commented:
I have no idea why this failed after I rebased it, and the errors are in NAGtypes of all places!

mahrud commented:
Figured it out, broken because of https://github.com/Macaulay2/M2/pull/3494#issuecomment-2451082376.

mahrud commented:
Argh this isn't merged, I just made a mistake. I'll reopen a new one.
